### ISSUE TITLE
remove rec from vic_run api

### ISSUE THREAD

**Issue by jhamman** [UW-Hydro/VIC#266]
The current `vic_run` api looks like this:

```C
int
vic_run(int                  rec,
        atmos_data_struct   *atmos,
        all_vars_struct     *all_vars,
        dmy_struct          *dmy,
        global_param_struct *gp,
        lake_con_struct     *lake_con,
        soil_con_struct     *soil_con,
        veg_con_struct      *veg_con,
        veg_lib_struct      *veg_lib,
        veg_hist_struct     *veg_hist)
```

The `rec` argument represents the current timestep and I'd like to argue it doesn't need to be passed into `vic_run`.  It ends up being passed throughout many `vic_run` functions but is only used twice:

```C
func_surf_energy_bal.c:    if (rec == 0) {
vic_run.c:    if (rec >= 0) {
```
In both cases, `rec` is used as a flag for initialization.  My argument is that in both of these cases, the code should be refactored.

In `vic_run.c`:
```C
    /* Convert LAI back to global */
    if (rec >= 0) {
        for (iveg = 0; iveg < Nveg; iveg++) {
            for (band = 0; band < Nbands; band++) {
                veg_var[iveg][band].LAI *= veg_var[iveg][band].vegcover;
                veg_var[iveg][band].Wdmax *= veg_var[iveg][band].vegcover;
            }
        }
    }
```
could become 
```C
    /* Convert LAI back to global */
    for (iveg = 0; iveg < Nveg; iveg++) {
        for (band = 0; band < Nbands; band++) {
            veg_var[iveg][band].LAI *= veg_var[iveg][band].vegcover;
            veg_var[iveg][band].Wdmax *= veg_var[iveg][band].vegcover;
        }
    }
```
since rec is always >= 0 in `vic_run`.

and in `func_surf_energy_bal.c`:
```C
    static int         error_cnt0, error_cnt1;

...

    if (rec == 0) {
        error_cnt0 = 0;
        error_cnt1 = 0;
    }
```
Could be refactored as 
```C
    static int         error_cnt0 = 0;
    static int         error_cnt1 = 0;

...

// removed initialization
```

I'm already in the process of only passing in the current timesteps `dmy_struct` and these two changes would go nicely together.

**Comment by jhamman** [comment]
I should add that keeping the `vic_run` api as clean as possible will make (is making) adding new drivers much easier.

**Comment by bartnijssen** [comment]
Isn't it used in the `veg_hist` part of the code. If I remember correctly, this was one of the reasons why I asked @tbohn to refactor this.

I completely agree otherwise.

**Comment by jhamman** [comment]
I thought so too but [this](https://github.com/UW-Hydro/VIC/blob/develop/vic/vic_run/src/vic_run.c#L277-L279) is the typicall usage of `rec` and `dmy` in `vic_run`.  

```C
displacement[0] = vic_run_veg_lib[pet_veg_class].displacement[dmy[rec].month - 1];
```

So by only passing in a single `dmy`, you can get rid of `rec` all together:
```C
displacement[0] = vic_run_veg_lib[pet_veg_class].displacement[dmy->month - 1];
```

**Comment by bartnijssen** [comment]
I think I already changed this at the time (originally `veg_hist` was passed in entirely). The next refactor is not to pass in `veg_hist` at all since the vegetation structures can be updated upstream of `vic_run`.

Please proceed as you suggest.

**Comment by jhamman** [comment]
Okay, a PR is forthcoming.

**Comment by jhamman** [comment]
closed via #267.
